admin-dev/themes/new-theme/js/pages/order/create/product-manager.js

```diff
@@ -0,0 +1,245 @@
+   * @private
+   */
+  _initEvents() {
+    $(createOrderMap.productSearch).on('input', event => this._search(event));
+    $(createOrderMap.productSelect).on('change', (event) => {
+      const productId = Number($(event.currentTarget).find(':selected').val());
+      this._selectProduct(productId);
+    });
+    $(createOrderMap.combinationsSelect).on('change', (event) => {
+      const combinationId = Number($(event.currentTarget).find(':selected').val());
+      this._selectCombination(combinationId);
+    });
+  }
+
+  /**
+   * Searches for product
+   *
+   * @private
+   */
+  _search(event) {
+    const minSearchPhraseLength = 3;
+    const $productSearchInput = $(event.currentTarget);
+    const name = $productSearchInput.val();
+
+    if (name.length < minSearchPhraseLength) {
+      return;
+    }
+
+    $.get(this.router.generate('admin_products_search'), {
+      search_phrase: name,
+    }).then((response) => {
+      this.products = JSON.parse(response);
+      this.renderer.renderSearchResults(this.products);
+      this._selectFirstResult();
+    }).catch((response) => {
+      if (typeof response.responseJSON !== 'undefined') {
+        showErrorMessage(response.responseJSON.message);
+      }
+    });
+  }
+
+  /**
+   * Initiate first result dataset after search
+   *
+   * @private
+   */
+  _selectFirstResult() {
+    this._unsetProduct();
+
+    if (this.products.length !== 0) {
+      this._selectProduct(Object.keys(this.products)[0]);
+    }
+  }
+
+  /**
+   * Handles use case when product is selected from search results
+   *
+   * @private
+   *
+   * @param productId
+   */
+  _selectProduct(productId) {
+    this._unsetCombination();
+
+    this.selectedProductId = productId;
+    const product = this.products[productId];
+
+    this.renderer.renderProductMetadata(product);
+
+    // if product has combinations select the first else leave it null
+    if (product.combinations.length !== 0) {
+      this._selectCombination(Object.keys(product.combinations)[0]);
+    }
+
+    return product;
+  }
+
+  /**
+   * Handles use case when new combination is selected
+   *
+   * @param combinationId
+   *
+   * @private
+   */
+  _selectCombination(combinationId) {
+    const combination = this.products[this.selectedProductId].combinations[combinationId];
+
+    this.selectedCombinationId = combinationId;
+    this.renderer.renderStock(combination.stock);
+
+    return combination;
+  }
+
+  /**
+   * Sets the selected combination id to null
+   *
+   * @private
+   */
+  _unsetCombination() {
+    this.selectedCombinationId = null;
+  }
+
+  /**
+   * Sets the selected product id to null
+   *
+   * @private
+   */
+  _unsetProduct() {
+    this.selectedProductId = null;
+  }
+
+  /**
+   * Adds selected product to current cart
+   *
+   * @private
+   */
+  _addProductToCart(cartId) {
+    this.cartEditor.addProduct(cartId, this._getProductData(cartId));
+    EventEmitter.on(eventMap.productAddedToCart, (cartInfo) => {
```

Comment:
The `addProduct` method emits a `productAddedToCart` event after the ajax request ends and then I do `something` on that event. Not sure what you mean by `unlistening` ? 😄

Reply:
Well events are like listeners, each time you call the `on` method you add a callback to the list of listeners for this method. But in this case each time `_addProductToCart` is called you add a new listener, so on second call you will have two listeners, on the third three listeners, and so on...
So the callback methods will be called multiple times each time you add a product. To avoid this you have different solutions:
- either you init your listener in a separate method and call it only *once*
- remove the listener after it's been called using removeListener but you need the reference on the callback so you can't use an arrow function, you will have to use a named class function
- in this case you are lucky because you only have one listener (no failure event) so you can use `EventEmitter.once` instead of `EventEmitter.on` It works the same way except it automatically removes the listener once it's called

The `once` can do the job here because you only have one event to listen and you need to be sure that it is *always* fired. But in other cases like in `_addCartRuleToCart` you listen to two events so it's not appropriate You have to init both events beforehand and only once Also be careful it's usually a good practice to clear the events when you stop using the objects because as long as they are present the object is kept in memory and the garbage collector can't remove it

Finally you should always init your listener before you call the method that emits them, if you're dealing with asynchronous methods you might be good But if the method is executed instantly the event will emitted before you register it and you won't be able to catch it any more